subprojects/antlr/src/main/java/org/gradle/api/plugins/antlr/AntlrPlugin.java

```diff
@@ -68,16 +81,20 @@ public void apply(final Project project) {
         // Wire the antlr configuration into all antlr tasks
         project.getTasks().withType(AntlrTask.class).configureEach(antlrTask -> antlrTask.getConventionMapping().map("antlrClasspath", () -> project.getConfigurations().getByName(ANTLR_CONFIGURATION_NAME)));
 
-        project.getExtensions().getByType(JavaPluginExtension.class).getSourceSets().all(
+        JavaPluginExtension javaPluginExtension = project.getExtensions().getByType(JavaPluginExtension.class);
+        javaPluginExtension.getSourceSets().all(
                 new Action<SourceSet>() {
                     @Override
                     public void execute(final SourceSet sourceSet) {
                         // for each source set we will:
                         // 1) Add a new 'antlr' virtual directory mapping
+
+
                         final AntlrSourceVirtualDirectoryImpl antlrDirectoryDelegate
-                                = new AntlrSourceVirtualDirectoryImpl(((DefaultSourceSet) sourceSet).getDisplayName(), objectFactory);
+                                = instantiator.newInstance(AntlrSourceVirtualDirectoryImpl.class, ((DefaultSourceSet) sourceSet).getDisplayName(), objectFactory, patternSetFactory, fileCollectionFactory, directoryFileTreeFactory);
```

Comment:
Can we use `objectFactory` here instead, and let it inject the services?

Reply:
Yep, I'll update that.
In fact, this is a WIP PR and I'll update it to match to what we have in https://github.com/gradle/gradle/pull/17149.